[Chng Ian] iP

chngchngchng opened the pull request:
### DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is __FREE__!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a  Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

chngchngchng committed:
Duke.java: Added functionality to echo user input for Level 1

chngchngchng committed:
Duke.java: Added functionality to add tasks to list and echo list for Level 2

chngchngchng committed:
Task.java, Duke.java: Implemented Level 3 Functionality

chngchngchng committed:
main/java/*: Added functionality for Level 3

chngchngchng committed:
main/java/*: Added functionality for Level 4

chngchngchng committed:
Merge branch 'master' of https://github.com/chngchngchng/cs2103t-ip

chngchngchng committed:
All files: Added basic testing for Level 5

chngchngchng committed:
Duke.java: Added delete functionality for Level 6

chngchngchng committed:
DukeParser.java, Duke.java: Refactor Duke code to include instruction parser

chngchngchng committed:
TaskList.java: Abstract list functionality into separate class

chngchngchng committed:
All files: Add save and load from disk functionality, repackaged classes

chngchngchng committed:
DukeParser.java: Added datetime functionality for deadlines and events

chngchngchng committed:
Fixed merge conflicts

chngchngchng committed:
Merge tag 'branch-Level-8'

chngchngchng committed:
Merge branch 'master' into branch-Level-8

chngchngchng committed:
DukeParser.java: Fixed small print statement bugs

chngchngchng committed:
Revert "DukeParser.java: Fixed small print statement bugs"

This reverts commit 9bfdeadb3f55e53eb936203e3c954b4d029312d0.

chngchngchng committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

chngchngchng committed:
*: Added Gradle dependencies to build and run project

chngchngchng committed:
src/test/*: Added basic testing using JUnit to Parser, TaskList

chngchngchng committed:
*: Added javadoc to all files

chngchngchng committed:
*: Fixed code style to comply with coding standard

chngchngchng committed:
*: Refactored JavaDoc comments to match conventions

chngchngchng committed:
DukeParser.java, TaskList.java: Added find functionality to search for user tasks

chngchngchng committed:
Merge branch 'A-CodingStandard'

chngchngchng committed:
Merge branch 'branch-Level-9'

chngchngchng committed:
*: Added checkstyle compatibility and fixed issues

chngchngchng committed:
*: Add GUI support, and refactored commands into individual classes

johnbenedictyan commented on src/main/java/duke/parser/DukeParser.java:
```diff
@@ -0,0 +1,235 @@
+    /**
+     * Executes a loaded and parsed instruction.
+     *
+     * @param st Storage that we want to save data to after the instruction has been executed
```
Perhaps this param was changed? Now, should this param tag be needed?

chngchngchng replied:
Yep, I should remove this param.

johnbenedictyan commented on src/main/java/duke/parser/DukeParser.java:
```diff
@@ -0,0 +1,235 @@
+        // First, try to parse the numerical part of the instruction, if error, throw it
+        int instructionNum;
+        try {
+            instructionNum = Integer.valueOf(this.restOfInputString) - 1;
```
Any reason why you used `Integer.valueOf` instead of `parseInt`?

chngchngchng replied:
Seemed to be the first option that came to mind! Is there any difference in efficiency in using one over the other?

johnbenedictyan replied:
`Integer.valueOf` actually uses `parseInt` internally, but it returns a type of Integer rather than int. I just thought for your case, `parseInt` was good enough. No major performance issues if I recall correctly.

```
public static int parseInt(String s) throws NumberFormatException {
    return parseInt(s, 10);
}

public static Integer valueOf(String s, int radix) throws NumberFormatException {
    return Integer.valueOf(parseInt(s, radix));
}

public static Integer valueOf(String s) throws NumberFormatException {
    return Integer.valueOf(parseInt(s, 10));
}
```

johnbenedictyan commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Override
+    public String toString() {
```
Perhaps you could use StringBuilder in this instance to help with performance
```
int counter = 1;
StringBuilder res = new StringBuilder("Here are the tasks that you have added to the list:\n");
for (Task task : this.taskList) {
    if (task != null) {
        res.append(counter).append(". ").append(task);
        if (counter != taskList.size()) {
            res.append("\n");
        }
        counter++;
    }
}
return counter == 1 ? "There are no tasks in your task list at the moment!" : res.toString();
```

chngchngchng replied:
Good idea! Thanks for your input.

leecaregene commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Duke project template
+# duke.Duke project template
```
Is this change unintentional?
```suggestion
# Duke project template
```

johnbenedictyan commented on src/main/java/duke/parser/DukeParser.java:
```diff
@@ -0,0 +1,235 @@
+
+import java.time.format.DateTimeParseException;
+
+import duke.commands.*;
```
Perhaps the imported classes should be listed explicitly?

johnbenedictyan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,56 @@
+package duke;
+
+import java.util.Scanner;
```
Maybe this unused import should be deleted, I noticed the same issue in other places too.

leecaregene commented on src/main/java/duke/parser/DukeParser.java:
```diff
@@ -0,0 +1,235 @@
+     *
+     * @param inputString The string that we would like to parse.
+     */
+    public void parseInput(String inputString)  {
```
Extra space?
```suggestion
    public void parseInput(String inputString) {
```

leecaregene commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,120 @@
+                String data = sc.nextLine();
+                String[] parsedData = data.split(",,");
+                if (parsedData.length != 4) {
+                    throw new DukeException("Error reading from file: " + filePath + ". Perhaps " +
+                            "the data is stored in an incorrect manner?");
+                }
+                String type = parsedData[0];
+                boolean isDone;
+                try {
+                    isDone = Boolean.parseBoolean(parsedData[1]);
+                } catch (Exception e) {
+                    throw new DukeException("Error reading from file: " + filePath + ": Unknown " +
+                            "boolean format for stored tasks");
+                }
+
+                // TODO: More error checking for description and datetimeinfo?
+                String description = parsedData[2];
+                String dateTimeInfo = parsedData[3];
+
+                switch (type) {
+                case "TODO":
+                    newList.add(new Todo(description, isDone));
+                    break;
+                case "DEADLINE":
+                    newList.add(new Deadline(description, isDone, dateTimeInfo));
+                    break;
+                case "EVENT":
+                    newList.add(new Event(description, isDone, dateTimeInfo));
+                    break;
+                default:
+                    throw new DukeException("Error reading from file: " + filePath + ": Unknown " +
+                            "task type");
```
Line break should be before an operator?
```suggestion
                    throw new DukeException("Error reading from file: " + filePath + ". Perhaps "
                    + "the data is stored in an incorrect manner?");
                }
                String type = parsedData[0];
                boolean isDone;
                try {
                    isDone = Boolean.parseBoolean(parsedData[1]);
                } catch (Exception e) {
                    throw new DukeException("Error reading from file: " + filePath + ": Unknown "
                    + "boolean format for stored tasks");
                }

                // TODO: More error checking for description and datetimeinfo?
                String description = parsedData[2];
                String dateTimeInfo = parsedData[3];

                switch (type) {
                case "TODO":
                    newList.add(new Todo(description, isDone));
                    break;
                case "DEADLINE":
                    newList.add(new Deadline(description, isDone, dateTimeInfo));
                    break;
                case "EVENT":
                    newList.add(new Event(description, isDone, dateTimeInfo));
                    break;
                default:
                    throw new DukeException("Error reading from file: " + filePath + ": Unknown "
                    + "task type");
```

leecaregene commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+        for (Task task: this.taskList) {
+            for (String term: terms) {
+                if (task.toString().toLowerCase().contains(term.toLowerCase())) {
+                   if (!result.contains(task.toString())) {
+                       result += task.toString() + "\n";
+                   }
```
Indentations off by one space?
```suggestion
                    if (!result.contains(task.toString())) {
                        result += task.toString() + "\n";
                    }
```

leecaregene commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,48 @@
+
+
+    public String getStatusIcon() {
+        return isDone? "X" : " ";
```
Missing space?
```suggestion
        return isDone ? "X" : " ";
```

leecaregene reviewed (commented):
Some coding standard violations, you can use checkstyle to find the rest.

chngchngchng committed:
DukeParser.java: Refactor code logic

Within DukeParser.java, the addTaskInstructionHandler has a lot of extra logic that can be handled

Abstracting out this functionality into its own function adheres to the SLAP principle and makes code cleaner, so let's do that.

chngchngchng committed:
DialogBox.fxml: Fix bug where label text in GUI truncates if length is too long

chngchngchng committed:
TaskList.java: Change toString method to make use of StringBuilder

StringBuilder is an inbuilt Java class that speeds up string
concatenation and building

Making use of it will ensure that code runs faster and cleaner, so let's
do that.

chngchngchng committed:
Duke.java, TaskList.java: Add assertions to code

Assertions help ensure that certain conditions are met when the code is
running. If these conditions are not met, then the code stops running.

Adding assertions to code helps to catch issues that may arise during
runtime, especially hard bugs. Let's add some to our code.

chngchngchng committed:
TaskList.java: Add more StringBuilder functionality

chngchngchng committed:
TaskList.java: Fix find functionality bugs

chngchngchng committed:
Merge pull request #1 from chngchngchng/branch-A-CodeQuality

Branch A-CodeQuality

chngchngchng committed:
Merge branch 'master' into branch-A-Assertions

chngchngchng committed:
Merge pull request #2 from chngchngchng/branch-A-Assertions

Duke.java, TaskList.java: Add assertions to code

chngchngchng committed:
*.java: Add reminder command and refactor tasks for better polymorphic behaviour

chngchngchng committed:
Merge branch 'branch-B-Reminders'

chngchngchng committed:
HelpCommand.java: Add help command to Duke for user help

chngchngchng committed:
DukeParser.java: Add apache commons cli parser for friendlier syntax

chngchngchng committed:
DialogBox.java: Add rounded edges to image

chngchngchng committed:
Merge branch 'branch-Level-10'

chngchngchng committed:
README.md: Add user guide for Duke
